File: AEDS2/Verde/WarmUp-Prova2/Q02/Bingo.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

public class Bingo {
    public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String entrada1;
        while((entrada1 = br.readLine()) != null) {
            String[] partes = entrada1.split(" ");
            int N = Integer.parseInt(partes[0]);
            int K = Integer.parseInt(partes[1]);
            int maiorNatural = Integer.parseInt(partes[2]);

            String entrada2;
            Cartela cartela = new Cartela(K, N);
            Cartela cartelaVencedora = new Cartela(maiorNatural);
            int tempo = maiorNatural + 1;
            int vencedor = -1;
            int contador = 0;
            
            for(int i = 0; i < N; i++) {
                entrada2 = br.readLine();
                cartela.inserir(entrada2, K);
                cartela.inserir(contador++);
            }

            entrada2 = br.readLine();
            cartelaVencedora.inserir(entrada2, maiorNatural);
                        
            for(int i = 0; i < N; i++) {
                if(cartela.completou(cartelaVencedora, i)) {
                    if(cartela.tempoCompletar(cartelaVencedora, i) < tempo) {
                        tempo = cartela.tempoCompletar(cartelaVencedora, i);
                        vencedor = i + 1;
                    }
                }
            }

            System.out.println(vencedor);
        }
    }


}

class Cartela {
    private int []array;
    private int [][]arrayCartelas;

    public Cartela(int K, int N) {
        this.array = new int[K];
        this.arrayCartelas = new int[N][K];
    }
    public Cartela(int K) {
        this.array = new int[K];
        this.arrayCartelas = null;
    }

    public void inserir(String entrada, int k) {
        String[] partes = entrada.split(" ");
        for(int i = 0; i < k; i++) {
            array[i] = Integer.parseInt(partes[i]);
        }
    }

    public void inserir(int contador) {
        for(int i = 0; i < array.length; i++) {
            arrayCartelas[contador][i] = array[i];
        }
    }

    public boolean completou(Cartela vencedora, int count) {
        int n = vencedora.array.length;
        int contador = 0;
        boolean numeroEncontrado;
        for(int i = 0; i < n; i++) {
            numeroEncontrado = false;
            for(int j = 0; j < array.length && !numeroEncontrado; j++) {
                if(arrayCartelas[count][j] == vencedora.array[i]) {
                    contador++;
                    numeroEncontrado = true;
                }
            }
        }
        if(contador == array.length) {
            return true;
        }
        else {
            return false;
        }
    }

    public int tempoCompletar(Cartela vencedora, int count) {
        int n = vencedora.array.length;
        int contador = 0;
        int tempo = 0;
        boolean numeroEncontrado;
        for(int i = 0; i < n && contador != array.length; i++) {
            numeroEncontrado = false;
            for(int j = 0; j < array.length && !numeroEncontrado && contador != array.length; j++) {
                if(arrayCartelas[count][j] == vencedora.array[i]) {
                    contador++;
                    numeroEncontrado = true;
                }
            }
            tempo++;
        
        }
        return tempo;
    }
}

File: AEDS2/Verde/WarmUp-Prova2/Q02/Bingo.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>


char **split(char* str, const char* delimitador) {
    int contador = 0;
    char* tmp = strdup(str);
    char* token = strtok(tmp, delimitador);
    while(token != NULL) {
        contador++;
        token = strtok(NULL, delimitador);
    }
    free(tmp);

    char** splits = (char**)malloc((contador + 1) * sizeof(char*));
    tmp = strdup(str);
    contador = 0;
    token = strtok(tmp, delimitador);

    while(token != NULL) {
        splits[contador] = strdup(token);
        contador++;
        token = strtok(NULL, delimitador);
    }
    splits[contador] = NULL;
    free(tmp);

    return splits;
}

int main() {
    char *str = (char*) malloc(250 * sizeof(char));
    char *entrada = (char*) malloc(250 * sizeof(char));

    while(scanf(" %[^\n]", str) != EOF) {
    char** partes = split(str, " ");
    int N = atoi(partes[0]);
    int K = atoi(partes[1]);
    int maiorNatural = atoi(partes[2]);
    int *cartela;
    int **cartelas = (int**) malloc(N * sizeof(cartela));
    for(int i = 0; i < N; i++) {
        scanf(" %[^\n]", entrada);
        partes = split(entrada, " ");
        cartela = (int*)malloc(K * sizeof(int));
        for(int j = 0; j < K; j++) {
            cartela[j] = atoi(partes[j]);
        }
        cartelas[i] = cartela;
        for(int j = 0;partes[j] != NULL; j++) {
            free(partes[j]);
        }
        free(partes);
    }

    scanf(" %[^\n]", entrada);
    partes = split(entrada, " ");
    int cartelaSorteada[maiorNatural];
    for(int i = 0; i < maiorNatural; i++) {
        cartelaSorteada[i] = atoi(partes[i]);
    }

    
    int contador = 0;
    int tempo = 0;
    int tempoVencedor = maiorNatural + 1;
    int vencedor = maiorNatural + 1;
    bool numeroEncontrado;
    for(int i = 0; i < N; i++) {
        for(int k = 0; k < maiorNatural && contador != K; k++) {
            numeroEncontrado = false;
            for(int j = 0; j < K && !numeroEncontrado && contador != K; j++) {
                if(cartelas[i][j] == cartelaSorteada[k]) {
                    contador++;
                    numeroEncontrado = true;
                }
            }

            tempo++;
        }
        
        if(tempo < tempoVencedor){
            tempoVencedor = tempo;
            vencedor = i+1;
        }
            
        
        contador = tempo = 0;
    }

    printf("%d\n", vencedor);
    }
    return 0;


}